core/src/main/java/org/apache/iceberg/util/StructLikeWrapper.java

```diff
@@ -80,12 +108,20 @@ public boolean equals(Object other) {
 
   @Override
   public int hashCode() {
-    int result = 97;
-    int len = struct.size();
-    result = 41 * result + len;
-    for (int i = 0; i < len; i += 1) {
-      result = 41 * result + Objects.hashCode(struct.get(i, Object.class));
+    if (hashCode == null) {
+      if (structHash != null) {
+        this.hashCode = structHash.hash(struct);
+      } else {
+        int result = 97;
```

Comment:
Is this the work-around when we don't have a type? In which case does this happen?

Reply:
Yes. This is a work-around for when there is no type, so that the current code paths don't fail. The follow-up commit removes this after adding `specId` to `DataFile`.